metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -0,0 +1,249 @@
+package util
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/kubernetes"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+// Enricher takes Kubernetes events and enrich them with k8s metadata
+type Enricher interface {
+	// Start will start the Kubernetes watcher on the first call, does nothing on the rest
+	// errors are logged as warning
+	Start()
+
+	// Stop will stop the Kubernetes watcher
+	Stop()
+
+	// Enrich the given list of events
+	Enrich([]common.MapStr)
+}
+
+type kubernetesConfig struct {
+	// AddMetadata enables enriching metricset events with metadata from the API server
+	AddMetadata bool          `config:"add_metadata"`
+	InCluster   bool          `config:"in_cluster"`
+	KubeConfig  string        `config:"kube_config"`
+	Host        string        `config:"host"`
+	SyncPeriod  time.Duration `config:"sync_period"`
+}
+
+type enricher struct {
+	sync.RWMutex
+	metadata       map[string]common.MapStr
+	index          func(common.MapStr) string
+	watcher        kubernetes.Watcher
+	watcherStarted bool
+}
+
+// GetWatcher initializes a kubernetes watcher with the given
+// scope (node or cluster), and resource type
+func GetWatcher(base mb.BaseMetricSet, resource kubernetes.Resource, nodeScope bool) (kubernetes.Watcher, error) {
+	config := kubernetesConfig{
+		AddMetadata: true,
+		InCluster:   true,
+	}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	// Return nil if metadata enriching is disabled:
+	if !config.AddMetadata {
+		return nil, nil
+	}
+
+	client, err := kubernetes.GetKubernetesClient(config.InCluster, config.KubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	options := kubernetes.WatchOptions{
+		SyncTimeout: config.SyncPeriod,
+	}
+
+	// Watch objects in the node only
+	if nodeScope {
+		options.Node = kubernetes.DiscoverKubernetesNode(config.Host, config.InCluster, client)
+	}
+
+	return kubernetes.NewWatcher(client, resource, options)
+}
+
+// NewResourceMetadataEnricher returns a Enricher configured for kubernetes resource events
+func NewResourceMetadataEnricher(
+	base mb.BaseMetricSet,
+	resource kubernetes.Resource,
+	nodeScope bool) Enricher {
+
+	watcher, err := GetWatcher(base, resource, nodeScope)
+	if err != nil {
+		logp.Warn("Error initializing Kubernetes metadata enricher: %s", err)
```

Comment:
I would prefer not to use `Warn` but either `Info` or `Error`. Seems more like `Info` in this case as even if the enricher fails, the metricset still starts?

Reply:
We tried to get rid of all Warn messages in the past. The reason is that for Warn it's not clear if users have to take again or not. It sounds like in the above case they have to take again so probably move it to `error`.